### PR TITLE
#4696: Build-Mode pattern now covers the whole reference

### LINKED ISSUE #4696: `--build <pattern>` seems not to work
When calling `conan help install` it states the following regarding the --build parameter:

```
--build=[pattern] Build always these packages from source, but never build the others. [...] 'pattern' is a fnmatch file pattern of a package name.
```

I would like to install a package and rebuild all of my own packages with it. All of my packages are within the `@monsdar/release` channel. The following seems not to work though: `conan install ... --build *monsdar*`. Conan gives me the message `ERROR: No package matching '*monsdar*' pattern`.

Do I do something wrong? Or is this feature not working correctly?

This happens with Conan 1.13.0 installed via Pip. Python 3.6.1 on a Windows 10 Professional with 64bit.

### PR DESCRIPTION
**Pull request by monsdar**
Fixes #4696

Changelog: Feature: `--build` parameter now applies fnmatching onto the whole reference, allowing to control rebuilding in a much broader way.
Docs: https://github.com/conan-io/docs/pull/XXXX

- [x] Refer to the issue that supports this Pull Request.
- [x] --If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.--
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

**Comment by monsdar** [comment]
Possibly breaking (but very unlikely):
When someone uses `conan install foo/1.0@user/stable --build B*ost`
* It was matching the dependency `Boost/1.69.0@user/host`
* It wasn't matching the dependency `Bar/1.0.0@user/host`

With the new implementation it will take the whole package reference an fnmatches it. Therefore the new implementation will match both references. This is a different behaviour like before.

The implementation works like expected with things like `Boost*` or similar though. It's just different when `*` is placed within the middle of the build_mode pattern.

**Comment by monsdar** [comment]
Not sure where the docs need to be expanded tbh... This feature is kind of cluttered throughout the documentation.

File: conans/client/graph/build_mode.py [hunk 1]
```diff
@@ -53,7 +53,10 @@ def forced(self, conan_file, ref):
 
         # Patterns to match, if package matches pattern, build is forced
         for pattern in self.patterns:
-            if fnmatch.fnmatch(ref.name, pattern) or repr(ref) == pattern:
+            is_matching_pattern = fnmatch.fnmatch(repr(ref), pattern)
+            is_exact_match = (repr(ref) == pattern)
+            is_same_component = repr(ref).startswith(pattern)
```
Comment (lasote) [inline conans/client/graph/build_mode.py]: Shouldn't this be also fnmatch (the old one actually)? That it means that if I use `--build=p` will work with any ref starting with `p`? Or am I missing something?

Comment (lasote) [reply conans/client/graph/build_mode.py]: Add an assert of a parameter without 🌟

**Comment by lasote** [comment]
We are going to need a PR for the docs also. Thanks!

File: conans/client/graph/build_mode.py (hunk 1 above)
Comment (monsdar) [reply conans/client/graph/build_mode.py]: Line 58 actually checks if the given reference starts with the pattern. So the check will return true in your case with `p`.
I can add this to the unittests if it isn't in there already.

Comment (monsdar) [reply conans/client/graph/build_mode.py]: Added a test that shows how this is still working. See `BuildModeTest::test_partly_build_force`

**Comment by monsdar** [comment]
PR for the docs: https://github.com/conan-io/docs/pull/1130

File: conans/client/graph/build_mode.py (hunk 1 above)
Comment (monsdar) [reply conans/client/graph/build_mode.py]: Added additional checks for parts of the name

**Comment by jgsogo** [comment]
I'm not sure about (examples taken from the tests) `"H"` matching `"Hey/0.1@user/testing"` or `"*Tool"` matching `"PythonTool/0.1@lasote/stable"`.... I would expect those patterns to match only for the exact `ref.name` (like it was before, I think), so:
 * `"Hey"` matches `"Hey/version@user/channel"` (no need for wildcards if it is equal to the `ref.name`)
 * `"*Tool"` matches `"name/version@user/whateverTool"`

But It is just an opinion, I haven't checked previous behavior...

**Comment by monsdar** [comment]
Supporting `*Tools` and `H` is the current behaviour. The unittest I added for `*Tools` is already part of the Jenkins tests.

**Comment by monsdar** [comment]
Jenkins still not passing all tests. Tbh I'm out of ideas what the test-output should tell me and how to make sure the tests pass. Could someone from your side help me out what the issue is and what needs to be done in order to have that solved?

**Comment by jgsogo** [comment]
On my way! 👍

**Comment by jgsogo** [comment]
The test that is failing is associated with one our comments above: "H" should not match "Hey/0.1@user/testing". I think that we only need two lines for this functionality to work:

```python
        for pattern in self.patterns:
            is_matching_name = fnmatch.fnmatch(ref.name, pattern)
            is_matching_ref = fnmatch.fnmatch(repr(ref), pattern)

            if is_matching_name or is_matching_ref:
                 ...
```

**Comment by monsdar** [comment]
Got it, that's easy to fix then

File: conans/client/graph/build_mode.py (hunk 1 above)
Comment (monsdar) [reply conans/client/graph/build_mode.py]: Removed the feature where `p` will start match `pattern` or `post` as discussed below. Not sure if I understood wrong from your message or what your intention was. Let me know when there's something still missing... So far it looks like all tests pass (added some more tests as well).

**Comment by monsdar** [comment]
Seems there's an issue with the MacOS runner... frozen for more than 90 minutes now...

**Comment by jgsogo** [comment]
I think now it is ready, I would only add a couple of tests trying to match with different casing... although I'm not sure if casing should be taken into account or not (ping @lasote)

**Comment by lasote** [comment]
I would keep the casing as it was (sensitive, right?)

**Comment by lasote** [comment]
@monsdar could you add a tiny test to verify that the case is important?

**Comment by lasote** [comment]
Np. we will add it. THANKS!

**Comment by monsdar** [comment]
Sorry, I haven't had time yet. With the implementation you merged it will check case-insensitive, so this is not working like it should. I changed the implementation to use `fnmatchcase` instead of `fnmatch` and added tests for that.

**Comment by monsdar** [comment]
Changed haven't been added, as the PR is already closed. I'll add a new one for the last changes. Feel free to merge that one as well if necessary.

**Pull request merged**